aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -68,7 +69,97 @@
 #include <numeric>
 #include <vector>
 
-namespace at { namespace native {
+namespace at {
+namespace meta {
+
+native::SCATTER_GATHER_OP get_operator_enum(const std::string& reduce) {
+  if (reduce == "add") {
+    return native::SCATTER_GATHER_OP::REDUCE_ADD;
+  } else if (reduce == "multiply") {
+    return native::SCATTER_GATHER_OP::REDUCE_MULTIPLY;
+  } else {
+    TORCH_CHECK(false, "reduce argument must be either add or multiply.");
+  }
+}
+
+template <typename Meta>
+void scatter_meta_impl(
+    Meta& meta,
+    const Tensor& self,
+    int64_t dim,
+    const Tensor& index,
+    const c10::optional<Tensor>& src = nullopt,
+    const c10::optional<std::string>& reduce = nullopt) {
+  int64_t wrapped_dim = at::maybe_wrap_dim(dim, self.dim());
+  at::native::scatter_gather_dtype_check("scatter", self, index, src);
+  at::native::scatter_shape_check(self, wrapped_dim, index, src);
+
+  if (src.has_value()) {
+    at::assert_no_overlap(self, src.value());
+  }
+
+  at::assert_no_internal_overlap(self);
+  at::assert_no_overlap(self, index);
```

Comment:
The overlap tests here look a little suspicious. Remember that you are writing the logic which will handle BOTH functional and inplace scatter variants. So if you have logic that doesn't test `maybe_get_output()`, then it needs to be universal in both cases.  Why is it necessary for self to not have internal overlap? Internal overlap is bad if we're scattering into the tensor: weird shit will happen. But if we're scattering into self *out of place*, overlap isn't a problem, since we're going to discard the overlap when we do the initial copy. We can see this in the old implementation too:

```
Tensor scatter(const Tensor &amp; self, int64_t dim, const Tensor &amp; index, const Scalar&amp; source) {  return self.clone(at::MemoryFormat::Preserve).scatter_(dim, index, source);
```

although `scatter_` is going to do the internal overlap test, it's always going to pass, because `clone()` has the invariant that it always gets rid of internal overlap.

Reply:
That makes sense. I hadn't really considered that. Thanks for pointing out!